examples/tracing.cpp

```diff
@@ -0,0 +1,153 @@
+
+// An example C++ type that stores an arbitrary JS value.
+struct SafeBox {
+  // Arbitrarily JS value that will need to be traced. The JS::Heap type has a
+  // constructor and destructor and will trigger any required GC barriers on
+  // read or write.
+  JS::Heap<JS::Value> stashed;
+
+  // If we provide a trace method, we can then be used in a JS::Rooted or
+  // similar and the GC will be able to successfuly trace us.
+  void trace(JSTracer* trc) { JS::TraceEdge(trc, &stashed, "stashed value"); }
+};
+
+static bool CustomTypeExample(JSContext* cx) {
+  // If we use SafeBox as a stack object, then a JS::Rooted is enough.
+  JS::Rooted<SafeBox> stackSafe(cx);
+
+  // We can also use js::UniquePtr if SafeBox should be allocated on heap.
+  JS::Rooted<js::UniquePtr<SafeBox>> heapSafe(cx, js::MakeUnique<SafeBox>());
+
+  // NOTE: A JS::Rooted<SafeBox*> is a compile error. If one wanted to support
+  // rooting bare non-GC pointers then both JS::MapTypeToRootKind and
+  // JS::GCPolicy need to be defined for SafeBox*. This should be avoided in
+  // favor of using a smart-pointer when possible.
+
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// By defining a JS::GCPolicy we can support tracing in existing types that we
+// cannot add a trace method to such as std::shared_ptr.
+//
+// In this example, we forward methods to GCPolicy of target type and provide
+// reasonable behaviours when we have no current target.
+template <typename T>
+struct JS::GCPolicy<std::shared_ptr<T>> {
+  static void trace(JSTracer* trc, std::shared_ptr<T>* tp, const char* name) {
+    if (T* target = tp->get()) {
+      GCPolicy<T>::trace(trc, target, name);
+    }
+  }
+  static bool needsSweep(std::shared_ptr<T>* tp) {
+    if (T* target = tp->get()) {
+      return GCPolicy<T>::needsSweep(target);
+    }
+    return false;
+  }
+  static bool isValid(const std::shared_ptr<T>& t) {
+    if (T* target = t.get()) {
+      return GCPolicy<T>::isValid(*target);
+    }
+    return true;
+  }
+};
+
+static bool ExistingTypeExample(JSContext* cx) {
+  // We can use std::shared_ptr too since we defined a GCPolicy above.
+  JS::Rooted<std::shared_ptr<SafeBox>> sharedSafe(cx,
+                                                  std::make_shared<SafeBox>());
+  return true;
+}
+
+////////////////////////////////////////////////////////////
+
+// When an embedding wishes to keep GC things alive even if the JavaScript no
+// longer cares has direct references, it must provide GC roots for the various
+// tracing mechanisms to search from. This is done using the PersistentRooted
+// type.
+//
+// Using a custom trace function can allow you to use a single PersistentRooted
+// to traverse the embeddings private state holding many GC pointers.
+
+// A global PersistentRooted is created before SpiderMonkey has initialized so
+// we must be careful that we do not create any JS::Heap fields during
```

Comment:
I'll let you get away with the passive voice here, but at least s/that we do not/to not/.